### PR TITLE
Make attorney assignment error actionable

### LINKED ISSUE #13110: Handle single failure in batch assignment of attorney tasks
As noted in [#13100](https://github.com/department-of-veterans-affairs/caseflow/pull/13100#issuecomment-572249616), now that we batch assign ama tasks, to handle when a single case assignment fails we should put the creation of all tasks inside of a transaction to fail when one fails. We should then also show an error that the user can act upon rather than the generic error.

## AC
- [ ] Front end reflects the tasks that have been assigned (and specifically not assigned)
   - [ ] Tasks move/stay in judge assign queue
   - [ ] Tasks counts are updated/stay the same
- [ ] Front end shows actionable error for the failed assignment

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves #13110

### Description
Pass and display the actual error to the front end and make some of our error messages more actionable. Also rename factory?

### Acceptance Criteria
- [x] See error generated by the back end on the front end

### Testing Plan
1. Log in as bvaaabshire and go to http://localhost:3000/queue/3/assign
2. Find an ama case and create a judge decision review task for it
```ruby
appeal = Appeal.find_by_uuid("<uuid>")
JudgeDecisionReviewTask.create!(appeal: appeal, assigned_to: User.third, parent: appeal.root_task)
```
3. Check case and an additional ama case and assign them both to an attorney
4. Confirm the error suggests the user refreshes the page, both cases remain checked, attorney case counts do not change, and the total number of cases does not change.
5. Confirm the above after refresh (minus the error of course)

### User Facing Changes
 - [x] More descriptive errors

 BEFORE|AFTER
 ---|---
Screen Shot 2020-04-29 at 4 39 20 PM|Screen Shot 2020-05-05 at 11 28 36 AM

### CODE REVIEW

File: app/controllers/judge_assign_tasks_controller.rb [hunk 1]
```diff
@@ -11,7 +11,7 @@ def create
   private
 
   def tasks
-    @tasks ||= ActiveRecord::Base.multi_transaction do
+    @tasks ||= ActiveRecord::Base.transaction do
```
Comment (hschallhorn) [inline app/controllers/judge_assign_tasks_controller.rb:14]: we are not touching [vacols](https://github.com/department-of-veterans-affairs/caseflow/blob/7ccac22775a6b8958bc8c1dbde10a164c86b6c7c/app/models/application_record.rb#L27), so this only needs to be transaction

File: app/models/task.rb [hunk 2]
```diff
@@ -226,8 +226,7 @@ def verify_org_task_unique
         Caseflow::Error::DuplicateOrgTask,
         appeal_id: appeal.id,
         task_type: self.class.name,
-        assignee_type: assigned_to.class.name,
-        parent_id: parent&.id
```
Comment (hschallhorn) [inline app/models/task.rb:230]: parent id was never used. axed

File: app/models/tasks/judge_decision_review_task.rb [hunk 3]
```diff
@@ -39,9 +39,8 @@ def verify_user_task_unique
       fail(
         Caseflow::Error::DuplicateUserTask,
         appeal_id: appeal.id,
-        task_type: self.class.name,
-        assignee_type: assigned_to.class.name,
-        parent_id: parent&.id
```
Comment (hschallhorn) [inline app/models/tasks/judge_decision_review_task.rb:44]: parent id was never used. axed

File: app/models/tasks/root_task.rb [hunk 4]
```diff
@@ -105,8 +105,7 @@ def verify_root_task_unique
         Caseflow::Error::DuplicateOrgTask,
         appeal_id: appeal.id,
         task_type: self.class.name,
-        assignee_type: assigned_to.class.name,
-        parent_id: parent&.id
```
Comment (hschallhorn) [inline app/models/tasks/root_task.rb:109]: parent id was never used. axed

File: client/app/queue/components/AssignWidget.jsx [hunk 5]
```diff
@@ -111,7 +112,7 @@ class AssignWidget extends React.PureComponent {
 
         return this.props.showErrorMessage({
           title: COPY.ASSIGN_WIDGET_ASSIGNMENT_ERROR_TITLE,
-          detail: errorDetail });
+          detail: errorMessages?.length ? errorMessages[0].title : errorDetail });
```
Comment (hschallhorn) [inline client/app/queue/components/AssignWidget.jsx:115]: Actually display the error if we have one

File: lib/caseflow/error.rb [hunk 6]
```diff
@@ -189,28 +189,31 @@ def initialize(args)
...
       @appeal_id = args[:appeal_id]
+      @veteran_id = args[:veteran_id]
       @task_type = args[:task_type]
       @assignee_type = args[:assignee_type]
       @code = args[:code] || 400
-      @message = args[:message] || "Appeal #{@appeal_id} already has an open task of type #{@task_type} assigned to "\
-                                   "#{assignee_type}. No action necessary"
+      @message = args[:message] || "Veteran's (#{@veteran_id}) appeal (#{@appeal_id}) already has an open task of type"\
+                                   " #{@task_type} assigned to #{assignee_type}. Please refresh the page. Contact "\
+                                   "support if this error persists."
```
Comment (hschallhorn) [inline lib/caseflow/error.rb:202]: Include veteran file number as it's useable by the user.

File: lib/caseflow/error.rb [hunk 7]
```diff
@@ -189,28 +189,31 @@ def initialize(args)
...
       @appeal_id = args[:appeal_id]
+      @veteran_id = args[:veteran_id]
       @task_type = args[:task_type]
-      @assignee_type = args[:assignee_type]
       @code = args[:code] || 400
-      @message = args[:message] || "Appeal #{@appeal_id} already has an open task of type #{@task_type} assigned to "\
-                                   "#{assignee_type}. No action necessary"
+      @message = args[:message] || "Veteran's (#{@veteran_id}) appeal (#{@appeal_id}) already has an open task of type"\
+                                   " #{@task_type} assigned to a user. Please refresh the page. Contact support if " \
+                                   "this error persists."
```
Comment (hschallhorn) [inline lib/caseflow/error.rb:216]: Include veteran file number as it's useable by the user. assignee_type is always "user" no need to pass it in

Comment (hschallhorn) [reply lib/caseflow/error.rb:216]: @geronimoramos How do ya feel about this wording?

 BEFORE|AFTER
 ---|---
Screen Shot 2020-04-29 at 4 39 20 PM|Screen Shot 2020-04-29 at 4 33 58 PM

Comment (geronimoramos) [reply lib/caseflow/error.rb:216]: @hschallhorn yes looks good! I would only change to:
1. 'already has an open task type of JudgeDecisionReviewTask assigned to a user' OR 
2. 'already has an open task of JudgeDecisionReviewTask JudgeDecisionReviewTask'

Do either of those make sense in this context? Also -- are we able to make 'Contact support if this error persists' an actionable link?

File: lib/caseflow/error.rb [hunk 8]
```diff
@@ -189,28 +189,31 @@ def initialize(args)
...
 
     def initialize(args)
       @appeal_id = args[:appeal_id]
+      @veteran_id = args[:veteran_id]
       @task_type = args[:task_type]
-      @assignee_type = args[:assignee_type]
       @code = args[:code] || 400
-      @message = args[:message] || "Appeal #{@appeal_id} already has an open task of type #{@task_type} assigned to "\
-                                   "#{assignee_type}. No action necessary"
+      @message = args[:message] || "Veteran's (#{@veteran_id}) appeal (#{@appeal_id}) already has an open task type of"\
```
Comment (yoomlam) [inline lib/caseflow/error.rb:214]: `appeal_id` isn't directly useful to the user. We should show the docket number, which is visible to the user on the page.

File: client/app/queue/components/AssignWidget.jsx [hunk 9]
```diff
@@ -100,9 +100,10 @@ class AssignWidget extends React.PureComponent {
             casePlural: pluralize('case', selectedTasks.length)
           })
         });
-      }, () => {
+      }, (error) => {
         this.props.resetSaveState();
 
+        const errorMessages = error.response.body?.errors;
```
Comment (yoomlam) [inline client/app/queue/components/AssignWidget.jsx:106]: Is `error` and `error.response` always non-null?

**Comment by yoomlam** [comment]
Somewhat inconsistent behavior:
1. When assigning several AMA appeals to an attorney, if one fails to be assigned, none of the appeals are assigned.
2. When assigning a mix of AMA and legacy appeals to an attorney, if one fails to be assigned, legacy appeals will be assigned but no AMA appeals are assigned.


Bug (I think): not sure what's going on but here's the procedure to replicate (hopefully):
1. Click on a legacy appeal along with an AMA appeal, then assign to attorney.

```ruby
                                ┌────────────────────────────────────────────────────────────────────────┐
LegacyAppeal 42 (legacy) ────── │ ID   │ STATUS    │ ASGN_BY     │ ASGN_TO     │ UPDATED_AT              │
└── RootTask                    │ 1837 │ on_hold   │             │ Bva         │ 2020-04-30 13:18:01 UTC │
    ├── JudgeAssignTask         │ 1838 │ completed │             │ BVAAABSHIRE │ 2020-05-01 22:35:41 UTC │
    └── JudgeDecisionReviewTask │ 1846 │ on_hold   │ BVAAABSHIRE │ BVAAABSHIRE │ 2020-05-01 22:35:41 UTC │
        └── AttorneyTask        │ 1847 │ assigned  │ BVAAABSHIRE │ BVALSHIELDS │ 2020-05-01 22:35:41 UTC │
                                └────────────────────────────────────────────────────────────────────────┘
=> nil
[35] pry(main)> leg.vacols_id
=> "3662856"
```

Legacy case continues to be shown even though `JudgeAssignTask` is completed:
image

Legacy case is also shown in atty's queue:
image

Refreshed case details page but it still shows a JudgeAssignTask:
image

Tested some other things, then noticed that it says 6 cases on the left, but showing only 1. This is fixed after a manual page refresh.
image

**Comment by hschallhorn** [comment]
@yoomlam There should not be any legacy appeals with judge tasks. You'll need to turn off the DAS feature toggle.

File: client/app/queue/components/AssignWidget.jsx (hunk 9 above)
Comment (hschallhorn) [reply client/app/queue/components/AssignWidget.jsx:106]: Yep!

File: lib/caseflow/error.rb (hunk 8 above)
Comment (hschallhorn) [reply lib/caseflow/error.rb:214]: Agreed!

File: spec/controllers/judge_assign_tasks_controller_spec.rb [hunk 10]
```diff
@@ -75,6 +77,19 @@
...
         it_behaves_like "attorney task assignment"
       end
+
+      context "when the assignment encounters an error" do
+        before { create(:ama_judge_decision_review_task, assigned_to: judge, parent: assign_tasks.last.parent) }
+
+        it "Returns an error and creates no tasks" do
+          subject
+
+          expect(response.status).to eq 400
```
[the remark this answers is not in the document]
Comment (hschallhorn) [reply spec/controllers/judge_assign_tasks_controller_spec.rb:87]: Screen Shot 2020-05-05 at 11 30 43 AM
Updated!

File: spec/controllers/judge_assign_tasks_controller_spec.rb [hunk 11]
```diff
@@ -10,7 +10,9 @@
     let!(:judge_staff) { create(:staff, :judge_role, sdomainid: judge.css_id) }
     let!(:second_judge_staff) { create(:staff, :judge_role, sdomainid: second_judge.css_id) }
 
-    let!(:assign_tasks) { Array.new(3) { create(:ama_judge_task, assigned_to: judge, parent: create(:root_task)) } }
+    let!(:assign_tasks) do
+      Array.new(3) { create(:ama_judge_assign_task, assigned_to: judge, parent: create(:root_task)) }
```
Comment (yoomlam) [inline spec/controllers/judge_assign_tasks_controller_spec.rb:14]: Do you mean to create 3 different RootTasks?

Comment (hschallhorn) [reply spec/controllers/judge_assign_tasks_controller_spec.rb:14]: Yep, if not, we'd get a duplicate task error

File: client/app/queue/components/AssignWidget.jsx [hunk 12]
```diff
@@ -100,14 +100,20 @@ class AssignWidget extends React.PureComponent {
...
+
+        try {
+          errorDetail = error.response.body.errors[0].detail;
+        } catch (ex) {
+          errorDetail = this.props.isModal && userId ?
+            <React.Fragment>
+              <Link to={`/queue/${userId}/assign`}>{COPY.ASSIGN_WIDGET_ASSIGNMENT_ERROR_DETAIL_MODAL_LINK}</Link>
+              {COPY.ASSIGN_WIDGET_ASSIGNMENT_ERROR_DETAIL_MODAL}
+            </React.Fragment> : COPY.ASSIGN_WIDGET_ASSIGNMENT_ERROR_DETAIL;
+        }
```
Comment (hschallhorn) [inline client/app/queue/components/AssignWidget.jsx:116]: Actually unpack the error we get back and display it to the user

**Pull request merged**